packages/gatsby/src/services/listen-to-webpack.ts

```diff
@@ -1,12 +1,18 @@
 import { Compiler } from "webpack"
 import { InvokeCallback } from "xstate"
 import reporter from "gatsby-cli/lib/reporter"
+import { getServerDataChanged } from "../utils/page-ssr-module/get-server-data-utils"
 
 export const createWebpackWatcher =
   (compiler: Compiler): InvokeCallback =>
   (callback): void => {
-    compiler.hooks.invalid.tap(`file invalidation`, file => {
+    compiler.hooks.invalid.tap(`file invalidation`, async file => {
       reporter.verbose(`Webpack file changed: ${file}`)
       callback({ type: `SOURCE_FILE_CHANGED`, file })
+
+      if (file && (await getServerDataChanged(file))) {
```

Comment:
As mentioned below I would remove this piece and add some extra logic inside SOURCE_FILE_CHANGED to see if the module is part of an SSR route, if it is then we mark as dirty anyways, even if it's unnecessary

Reply:
What do you mean by "inside SOURCE_FILE_CHANGED"? This is just an event sent to xstate.

Reply:
SOURCE_FILE_CHANGED does trigger an action or a service. If we can add logic inside the service or action to check if the file was actually part of an SSR page. If it is then we rebuild.

Reply:
> If we can add logic inside the service or action to check if the file was actually part of an SSR page

While not directly in the scope of this PR but what if a user just added `getServerData`? So the file was not a part of an SSR page before but now it is?

Also, our handling of this event is pretty convoluted in the state machine now, so it is not obvious how to wire it up into it without a major refactor:

`SOURCE_FILE_CHANGED` [transitions to `aggregatingFileChanges`](https://github.com/gatsbyjs/gatsby/blob/fc84b66b7329e8377d39c3910ef8e406738060d6/packages/gatsby/src/state-machines/waiting/index.ts#L47-L49) which in turn [generates `extractQueries` action on timeout](https://github.com/gatsbyjs/gatsby/blob/fc84b66b7329e8377d39c3910ef8e406738060d6/packages/gatsby/src/state-machines/waiting/index.ts#L56-L62)  which in turn starts [runQueries service](https://github.com/gatsbyjs/gatsby/blob/fc84b66b7329e8377d39c3910ef8e406738060d6/packages/gatsby/src/state-machines/develop/index.ts#L243-L245) which in turn transition to [`recompiling`](https://github.com/gatsbyjs/gatsby/blob/fc84b66b7329e8377d39c3910ef8e406738060d6/packages/gatsby/src/state-machines/develop/index.ts#L170-L176) on `done` (and that's where we need the data if file was SSR). But after `aggregatingFileChanges` the list of changed files is actually lost because it doesn't save any payload.

But even if it did save it, webpack has its own file aggregation but doesn't expose the full list of aggregated files to us. So we can't even rely on `invalid` hook from webpack which Michal correctly pointed in https://github.com/gatsbyjs/gatsby/pull/33398#issuecomment-932749451

My current plan is to address Michal's comment by triggering this action from `extractQueries` logic which literally re-reads all files on every source file change (I assume exactly because of this problem with webpack).

Reply:
I don't know if something we can rely on, but our watcher for query extractions also emit `SOURCE_FILE_CHANGED` - https://github.com/gatsbyjs/gatsby/blob/eed610813da645356316826a4558640ecc4365b5/packages/gatsby/src/query/query-watcher.ts#L137-L150

So maybe we can rely on this action, even tho webpack itself won't emit `invalid` events for more than 1 file?

Reply:
Yeah, also looking at it now

Reply:
I'm fine with only fixing it for the current component and re-group and fix it for nested ones. This PR already has the major use case working.

Reply:
Simplified the logic in https://github.com/gatsbyjs/gatsby/pull/33398/commits/09490eca99fd293c71395187bdc090549f066091